### DISCUSSION TITLE
Grain reconstruction errors on large datasets

### DISCUSSION THREAD

**Discussion by T0tmo** [mtex-toolbox/mtex#2280] [Ask Anything]
Hello everyone, 

I'm running into a poblem when calculating grains on large EBSD datasets. The maps were constructed with smaller maps on AZtec and are not perfectly rectangular, the boundaries of the map present a staircase effect outside of which are non-indexed points.  When using calcGrains, points seem to appear oustide the borders of the map and generate errors during the reconstruction. I'm using MTEX 5.11.2 and this is the process I go through : 
```
ebsd = ebsd('indexed');

ebsd = ebsd(ebsd.mad < mad_threshold);

[grains, ebsd.grainId, ebsd.mis2mean] = calcGrains(ebsd, 'angle', threshold0 * degree);

ebsd(grains(grains.grainSize < 2)) = [];

[grains, ebsd.grainId, ebsd.mis2mean] = calcGrains(ebsd, 'angle', threshold1 * degree);
```

This is the kind of error I get at the borders when calculating the grains : 

<img width="851" alt="Capture d’écran 2024-12-17 à 10 44 33" src="https://github.com/user-attachments/assets/9c4a09ad-ee05-428d-800e-8d5b69a64d79" />

And when applying the 'boundary', 'tight' option, lines appear across the whole map : 

<img width="593" alt="Capture d’écran 2024-12-17 à 10 42 49" src="https://github.com/user-attachments/assets/56761a6b-78fd-46e6-b04c-8aae73c2ef7e" />

One solution I have found that works on some datasets is cropping the maps to a rectangle to avoid the staircase effect on the boundaries. However, on this specific dataset cropping doesn't work. And the error strangely arrises only when I use the 'boundary', 'tight' option, when I don't use it, the borders are fine. This is what the full map looks like in that case :

![untitled](https://github.com/user-attachments/assets/0f085fa8-e053-4f35-b85c-fd25eab6eb28)

So I am a bit at a Ioss on why this is happening. If anyone has an explanation for the reason behind this kind of bug or knows of a method to solve it, I'd be glad to read it. 
Here is a link to the data : [Link to cpr/crc](https://drive.google.com/drive/folders/1HGuEA-bbxR_fGtsO5W9oc1t4X1rZ5yxG?usp=sharing)

Thanks in advance for any help provided.

Cheers,

Thomas

**Comment by zmichels** [comment]
I'm not sure if it helps, as I did it a little differently... but... 
I imported the data you linked from the CRC and tried the code below to compute grains. 
It generated the attached plot. 
I also used 5.11.2.

```matlab
% initial grain constuction:
% trying "tight" boundary, with "indexed" (no other filtering)
[grains,ebsd.grainId,ebsd.mis2mean]  = calcGrains(ebsd('indexed'),'threshold',[1 10]*degree,'boundary','tight');

% remove small grains from grainset...
grains = grains(grains.grainSize>2);
% ... and from EBSD set
ebsd = ebsd(grains);

% reconstruct grain boundaries again:
[grains,ebsd.grainId,ebsd.mis2mean]  = calcGrains(ebsd('indexed'),'threshold',[1 10]*degree,'boundary','tight');

% smooth boundaries
grains = smooth(grains,5);
```
![Mg3_250C_15_grains_GOS](https://github.com/user-attachments/assets/5f00c316-de6b-4541-87d3-0e7d0a52762f)

**Reply by kilir** [reply]
Hi,
in addition, you can play with the 'tight' parameter.
Using mtex 6, you can can skip one reconstruction iteration using the 'minpix' parameter.
```Matlab
grains  =ebsd('indexed').calcGrains('minpix',3,'boundary','tight',0.9)
```
![out](https://github.com/user-attachments/assets/ff11814d-d327-4526-8a54-5af128f86cc1)

Cheers,
Rüdiger


It looks like the stitching doesn't work too well everywhere.

**Reply by T0tmo** [reply]
Thanks a lot for your answers, they were a real help !

**Comment by zmichels** [comment]
Unrelated: Was this sample deformed in tension by any chance?

**Reply by T0tmo** [reply]
Indeed it was. But we haven't quite figured it out yet but we think those bands visible on the microstructure were due to polishing or machining of the initial part ... your thoughts on it are welcome.

**Reply by zmichels** [reply]
I asked based on the intragranular orientation dispersion patterns.
The bands sound fun to ponder given they aren't yet fully explained. I'll let you know if I think of anything to suggest.

Cheers.

**Reply by zmichels** [reply]
Is the tension/extension direction parallel to X-axis in map?

Is there any chance the starting material had a layered construction? 
Context: I've seen some similar grain-size layering in olivine samples that were prepared by progressively cold-pressing adding identical micro volumes prior to hot-pressing(?)... to make the starting material. Then those layers can in some deformations rotate like strain markers.


Some fun bits I plotted along the way...

Grain long-axes (for grains with aspect ratio > 1.5):
![Mg3_250C_15_magnesium_SPO_rose](https://github.com/user-attachments/assets/c10adb2c-ad0a-4c3f-b3f3-ac772fb26eb3)


eigenvectors (columns) of grain-scale ODTs (all possible grains) in specimen (top row) and crystal (bottom row) coordinates:
![Mg3_250C_15_magnesium_CVA_all_eigs](https://github.com/user-attachments/assets/b4810943-d743-4a6b-a0d4-7da64464bb7c)

grains colored by diameter:
![Mg3_250C_15_map_grains_by_eqDiam](https://github.com/user-attachments/assets/e2745eb6-1e2f-4efa-a96d-1f042133d4f9)
